services/PackageReader.ts

```diff
@@ -0,0 +1,163 @@
+  process.env.PACKAGE_STATUS_FILE || path.resolve(".", "packages.status");
+
+const packageFileContent = fs.readFile(packageStatusFile).then((doc) => {
+  console.log("Reading");
+  return doc.toString();
+});
+
+const searchableField: (fieldName: string) => string = (fieldName) =>
+  fieldName + ": ";
+
+const readSimpleFieldFromParagraph: (
+  packageParagraph: string,
+  fieldName: string
+) => string = (packageParagraph, fieldName) => {
+  const _fieldName = searchableField(fieldName);
+
+  const fieldValue = packageParagraph
+    .split(fieldLineDelimiter)
+    .filter((line) => line.startsWith(_fieldName))
+    .map((line) => line.replace(_fieldName, ""));
+  if (fieldValue) {
+    return fieldValue[0];
+  } else {
+    return "";
+  }
+};
+
+const readMultiLineFieldFromParagraph: (
+  packageParagraph: string,
+  fieldName: string
+) => string = (packageParagraph, fieldName) => {
+  const fieldValue = searchableField(fieldName);
+  var regex = new RegExp(fieldValue + "[\\s\\S]*?(\\n\\S)", "m");
+  var result = packageParagraph.match(regex);
+  if (result) {
+    return result[0].replace(result[1], "").replace(fieldValue, "");
+  } else {
+    return "";
+  }
+};
+
+const readRelationshipFieldFromParagraph: (
+  packageParagraph: string,
+  fieldName: string
+) => string[] = (packageParagraph, fieldName) => {
+  const fieldValue = readSimpleFieldFromParagraph(packageParagraph, fieldName);
+  if (fieldValue) {
+    return fieldValue
+      .toString()
+      .replace(/ *\([^)]*\) */g, "")
+      .split(", ");
+  } else {
+    return [];
+  }
+};
+
+const readReverseRelationshipField: (
+  document: string,
+  fieldName: string,
+  packageName: string
+) => string[] = (document, fieldName, packageName) => {
+  const readRelationshipField = (paragraph: string) =>
+    readRelationshipFieldFromParagraph(paragraph, fieldName);
+  const readSimpleField = (paragraph: string) =>
+    readSimpleFieldFromParagraph(paragraph, packageField);
+
+  const paragraphs = document.split(paragraphDelimiter);
+  return paragraphs
+    .filter((p) =>
+      readRelationshipField(p)?.some((pckg) => pckg === packageName)
+    )
+    .map((p) => readSimpleField(p)!);
+};
+
+export const listPackageNames: () => Promise<
+  { name: string; version: string }[]
+> = () => {
+  return packageFileContent.then((doc) =>
+    doc
+      .split(paragraphDelimiter)
+      .filter(
+        (paragraph) =>
+          readSimpleFieldFromParagraph(paragraph, packageField) !== undefined
+      )
+      .map((paragraph) => {
+        const packageName = readSimpleFieldFromParagraph(
+          paragraph,
+          packageField
+        )!;
+        const version = readSimpleFieldFromParagraph(paragraph, versionField)!;
+        return { name: packageName, version: version };
+      })
+  );
+};
+
+type PackageDetails = {
+  name: string;
+  version: string;
+  description?: string;
+  depends?: string[];
+  dependant?: string[];
+};
+
+export const readPackageDetails: (
```

Comment:
Suggestion or something to consider: after splitting splitting the contents into `paragraph`, the helper functions like `readSimpleFieldFromParagraph` use _filter_ to search a specific field from the `paragraph` which does work just fine 👍 

However, Javascripts' JSON object is a very good tool in cases like these where you want to find a some specific value from a set of data. It could be a good exercise to try out how the code would look if it would:
1. split the contents into `paragraph` like currently
2. map each `paragraph` into an Paragraph object by mapping each line into a property:
   "Package: libws-commons-util-java" becomes { package: "libws-commons-util-java" }
   Mapping can be done in two steps (I think):
   1. split all lines with `const [field, value] = line.split(': ');
   2. after splitting, multiline lines have not been split (`field` has value, `value` is undefined)
   3. using a `reducer` multilines lines should be appended to the `value` of the previous `field`
3. properties from the Paragraph object can now be mapped into returned properties like `depends` and `dependant`

Reply:
@Klaudeta here's an example of what I mean:

```
type Package = {
  package: string;
  description: string;
};

const mapPackage = (paragraph: string): Package => {
  const lines = paragraph.split(fieldLineDelimiter);

  const pkg = lines.reduce(
    (previousValue, currentValue) => {
      const [field, value] = currentValue.split(": ");

      // we can assume that Description is the only multiline field
      // so all lines that can't be split with ":" are Description
      return value
        ? { ...previousValue, [field.toLowerCase()]: value }
        : { ...previousValue, description: previousValue.description + field };
    },
    { description: "" } as Package
  );

  return pkg;
};

```

Reply:
I admit it might take me a while to get to this short-hand manipulation of the content all by myself. I understand the logic once I see it though. 
One flaw I notice here is that the code above assumes that the multiline description field with not contain any ":" inside its content which based on the rules is not the case. If this happens then the splitting of the Description field above will go wrong I think.